src/lib/cloud/digitalocean/provisioner.ts

```diff
@@ -33,13 +31,89 @@ const REDIRECT_URIS: [string] = [
   //  'http://localhost:10101'
 ];
 
+const STORAGE_KEY_OAUTH = 'DigitalOcean-OAuth';
+
 interface KeyPair {
   private: string;
   public: string;
 }
 
 class Provisioner {
-  constructor(private dispatch_: Function, private state_: any = {}) {}
+  constructor(
+    private sshKeyPair_ :any = {},
+    private network_ :any = {},
+    private storage_ = freedom['core.storage'](),
+
+    // TODO: give this concept a better name, or use a different abstraction?
+    private cloud_ :any = null,
+
+    // doOAuth_() may be called multiple times concurrently before an OAuth
+    // flow has been completed. In order to ensure that each call doesn't
+    // trigger an additional OAuth flow in a new tab, the promise returned by
+    // the first call is cached in this field, and this cached promise is
+    // returned for subsequent calls.
+    private promiseOAuth_ :Promise<Object> = null
+  ) {}
+
+  /*
+   * Returns whether a DO OAuth token is saved in storage.
+   */
+  public hasOAuth = () : Promise<boolean> => {
+    return new Promise((F, R) => {
+      this.getOAuthFromStorage_().then(
```

Comment:
Hm, I'm not following here - where does `F` come from in your version, without a wrapping promise to provide it? I thought we needed a wrapping promise here so we can fulfill with the wrapped promise's result converted to a boolean. I also don't follow "no need for an explicit catch handler" - do you mean the onRejected callback supplied as the second argument to `then()` (which is not the same thing as supplying a callback to `.catch()`)? (`getOAuthFromStorage_()` doesn't throw, it rejects, so I thought you'd need an onRejected callback to handle that case.) Sorry if I'm misunderstanding!